File: Graph/3532.Path_Existence_Queries_in_a_Graph_I/Python/3532.Path_Existence_Queries_in_a_Graph_I.py
class Solution(object):
    def pathExistenceQueries(self, n, nums, maxDiff, queries):
        """
        :type n: int
        :type nums: List[int]
        :type maxDiff: int
        :type queries: List[List[int]]
        :rtype: List[bool]
        """
        #=====================================#
        # Disjoint-set based traversal method #
        #=====================================#

        ############
        #Initialize
        ##### Length of arrays (i.e. nums, queries) #####
        len_nums, len_queries = len(nums), len(queries)

        ##### Record summary counter #####
        record_cnter = 0

        ##### Record array #####
        record_arry = [0 for _ in range(len_nums)]

        ##### Result array #####
        res_arry = [False for _ in range(len_queries)]


        ########################################################
        #Disjoint-set based loop traversal with recorded arrays
        ##### Step 1: Record indexed-values informations with array #####
        for nums_idx in range(1, len_nums):

            ##### Check if the current difference matched conditions or not #####
            if ((nums[nums_idx] - nums[(nums_idx - 1)]) > maxDiff):
                record_cnter += 1 #Keep updating/accumulating
            else:
                pass

            record_arry[nums_idx] = record_cnter #Keep updating/overwriting

        ##### Step 2: Looped-traversal with recorded array #####
        for queries_idx in range(len_queries):

            ##### Check if the current summary counter matched conditions or not #####
            if (record_arry[(queries[queries_idx])[0]] == record_arry[(queries[queries_idx])[1]]):
                res_arry[queries_idx] = True #Keep updating/overwriting
            else:
                pass

        return res_arry

File: Graph/3532.Path_Existence_Queries_in_a_Graph_I/C++/3532.Path_Existence_Queries_in_a_Graph_I.cpp
class Solution {
public:
    vector<bool> pathExistenceQueries(int n, vector<int> & nums, int maxDiff, vector<vector<int>> & queries) 
    {
        //===============================================//
        // Input type:                                   //
        //   - int (n-th nodes)                          //
        //   - vector<int> ref. (nums vector)            //
        //   - int (maximun difference value)            //
        //   - vector<vector<int>> ref. (queries vector) //
        // Return type:                                  //
        //   - vector<bool> (result vector)              //
        //===============================================//

        /*Initialize*/
        ///// Length of vectors (i.e. nums, queries) /////
        int len_nums = nums.size(), len_queries = queries.size();

        ///// Record vector /////
        vector<int> record_vect(len_nums, 0);

        ///// Record summary counter /////
        int record_cnter = 0;

        ///// Result vector /////
        vector<bool> res_vect(len_queries, false);


        /*Disjoint-set based loop traversal with recorded vectors*/
        ///// Step 1: Record indexed-values informations with vector /////
        for (int nums_idx = 1; (nums_idx < len_nums); (nums_idx++)) //Whole
        {
            ///// Check if the current difference matched conditions or not /////
            if ((nums[nums_idx] - nums[(nums_idx - 1)]) > maxDiff) { (record_cnter++); } //Keep updating, accumulating
            else { ; }

            record_vect[nums_idx] = record_cnter; //Keep updating, overwriting

        } //Whole

        ///// Step 2: Looped-traversal with recorded vector /////
        for (int queries_idx = 0; (queries_idx < len_queries); (queries_idx++)) //Whole
        {
            ///// Check if the current summary counter matched conditions or not /////
            if (record_vect[(queries[queries_idx])[0]] == record_vect[(queries[queries_idx])[1]])
            {
                res_vect[queries_idx] = true; //Keep updating, overwriting
            }
            else { ; }

        } //Whole

        return res_vect;
    }
};
